### PR TITLE
Updating the header file with DNS rules during policy updates

### PR DESCRIPTION
**Pull request by vipul-21**
Writing the DNS rules to the endpoint header/json file as we apply the policy changes, This keeps the `ep_config.json` and `ep_config.h` in sync with the current rules being applied the policy.

### CODE REVIEW

File: pkg/proxy/dns.go [hunk 1]
```diff
@@ -48,6 +48,8 @@ func (dr *dnsRedirect) setRules(wg *completion.WaitGroup, newRules policy.L7Data
 	}
 	dr.currentRules = copyRules(dr.redirect.rules)
 
+	log.Infof("Updating the DNS rules for endpoint %d", dr.redirect.endpointID)
+	dr.redirect.localEndpoint.SyncEndpointHeaderFile()
```
Comment (christarazi) [inline pkg/proxy/dns.go:52]: `SyncEndpointHeaderFile` is already called within the endpoint pkg (`pkg/endpoint`) and during policy recalculation. Why is this necessary here especially expanding the scope of the proxy to reach into the management of endpoints?

**Review by christarazi — COMMENTED** [review]
Thanks for the PR. I have a comment below on the approach.

File: pkg/proxy/dns.go (hunk 1 above)
Comment (vipul-21) [reply pkg/proxy/dns.go:52]: Hey sorry for the delay in response. So the intention here is to update the DNS rules to the `ep_config.json` when we apply the policy.
Afaik, it does not write the dns rules to the file when the policy is applied. So I was calling this function which is triggered when we want to write the dns rules to the file. 
Current implementation, I see that the data is written to the fs after the policy applied, but dns rules are not. Is there another way we can incorporate getting the DNS rules to be written to the fs ?

Comment (christarazi) [reply pkg/proxy/dns.go:52]: > Current implementation, I see that the data is written to the fs after the policy applied, but dns rules are not. Is there another way we can incorporate getting the DNS rules to be written to the fs ?

I'm not sure I understand the problem you are describing. The DNS rules are in fact written to the fs as they are fetched from the same function: https://github.com/cilium/cilium/blob/325b5df02179d3ae9c62d074aadd07636346f1fb/pkg/endpoint/endpoint.go#L2438

Comment (christarazi) [reply pkg/proxy/dns.go:52]: Also another note to keep in mind here, see the comment regarding locks and deadlocks on the line above of the linked code: this PR might further complicate the locking pattern by now having the proxy reach into the headerfile.

Comment (squeed) [reply pkg/proxy/dns.go:52]: I agree; this is probably not necessary here. Rather, we should be syncing the endpoint header file explicitly on regeneration, which IIRC is not done now.

**Comment by squeed** [comment]
Ah, right, I remember what the problem is here: the cached DNS rules include the IP of the **dns server**, a.k.a the peer pod, which can change at any point in time, independent of the lifecycle of the subject endpoint.

So, DNS rules can be out-of-date even when no endpoint-specific edges have been missed. This is a bummer.

The solution, IMO, is to replace the Trigger with a Controller that runs periodically. That way we can be reasonably sure that the set of rules in question is up-to-date.

**Comment by vipul-21** [comment]
> Ah, right, I remember what the problem is here: the cached DNS rules include the IP of the **dns server**, a.k.a the peer pod, which can change at any point in time, independent of the lifecycle of the subject endpoint.
> 
> So, DNS rules can be out-of-date even when no endpoint-specific edges have been missed. This is a bummer.
> 
> The solution, IMO, is to replace the Trigger with a Controller that runs periodically. That way we can be reasonably sure that the set of rules in question is up-to-date.

So do you also recommend calling the controller when there are new rules being added Or solely rely on the periodic update ? I guess there can be a scenario where rules are applied but state in fs is not updated as controller is yet to run. I think that's what you meant by ` reasonably sure` part.
How frequent can/should the controller run ? Maybe have that configurable too.

**Comment by christarazi** [comment]
> Ah, right, I remember what the problem is here: the cached DNS rules include the IP of the **dns server**, a.k.a the peer pod, which can change at any point in time, independent of the lifecycle of the subject endpoint.
> 
> So, DNS rules can be out-of-date even when no endpoint-specific edges have been missed. This is a bummer.
> 
> The solution, IMO, is to replace the Trigger with a Controller that runs periodically. That way we can be reasonably sure that the set of rules in question is up-to-date.

Writing the DNS rules to disk is ultimately best effort because it's done synchronously and if the pod is deleted in the meantime before the latest rules can be synced to disk, then 

> > Ah, right, I remember what the problem is here: the cached DNS rules include the IP of the **dns server**, a.k.a the peer pod, which can change at any point in time, independent of the lifecycle of the subject endpoint.
> > So, DNS rules can be out-of-date even when no endpoint-specific edges have been missed. This is a bummer.
> > The solution, IMO, is to replace the Trigger with a Controller that runs periodically. That way we can be reasonably sure that the set of rules in question is up-to-date.
> 
> So do you also recommend calling the controller when there are new rules being added Or solely rely on the periodic update ? I guess there can be a scenario where rules are applied but state in fs is not updated as controller is yet to run. I think that's what you meant by ` reasonably sure` part. How frequent can/should the controller run ? Maybe have that configurable too.

I'm still not sure I understand the problem being solved here.

Basically, `SyncEndpointHeaderFile()` is called each time the endpoint makes a DNS request. `SyncEndpointHeaderFile()` triggers code with a 5s interval meaning the sync to the fs only occurs once every 5s. 

Therefore, I don't yet see the arguement for why we should sync DNS rules to the fs on policy calculation (endpoint regeneration) when the code already syncs the DNS rules when there's a DNS request made, once every 5s.

**Comment by squeed** [comment]
@christarazi @vipul-21 

> Therefore, I don't yet see the arguement for why we should sync DNS rules to the fs on policy calculation (endpoint regeneration) when the code already syncs the DNS rules when there's a DNS request made, once every 5s.

There are two things cached in the endpoint directory for fast recovery on startup:
1. The list of domain names + IPs seen by this endpoint
2. The list of ports + dns server pod IPs that are allowed

No. 1 cannot change without a DNS request, so it makes sense that this is updated per-request. However, no. 2 has a completely unrelated lifecycle. In fact, the lifecycle may be completely independent of the node itself, as CoreDNS being rescheduled on a remote node requires this to be updated.

So, that's why we should periodically refresh the cached DNS information (a.k.a "write the header file"). Either we build a mechanism to somehow propagate dns rules selector updates (hard!), or we just periodically update endpoint cached DNS info.

Make sense?

**Comment by christarazi** [comment]
> @christarazi @vipul-21
> 
> > Therefore, I don't yet see the arguement for why we should sync DNS rules to the fs on policy calculation (endpoint regeneration) when the code already syncs the DNS rules when there's a DNS request made, once every 5s.
> 
> There are two things cached in the endpoint directory for fast recovery on startup:
> 
>     1. The list of domain names + IPs seen by this endpoint
> 
>     2. The list of ports + dns server pod IPs that are allowed
> 
> 
> No. 1 cannot change without a DNS request, so it makes sense that this is updated per-request. However, no. 2 has a completely unrelated lifecycle. In fact, the lifecycle may be completely independent of the node itself, as CoreDNS being rescheduled on a remote node requires this to be updated.
> 
> So, that's why we should periodically refresh the cached DNS information (a.k.a "write the header file"). Either we build a mechanism to somehow propagate dns rules selector updates (hard!), or we just periodically update endpoint cached DNS info.
> 
> Make sense?

That clarifies things a bit, thanks. What is the consequence if we don't update (2) on a periodic basis? I understand the rules are "out of date / out of sync" on the fs, but what is the actual consequence of them being out of date? Does the list of ports + DNS server pods IPs (coredns) never get updated in the endpoint caches?

**Comment by vipul-21** [comment]
@christarazi 
>  What is the consequence if we don't update (2) on a periodic basis?

As of now there shouldn't be any consequence, but this was discussed in sig-policy meeting in June: https://docs.google.com/document/d/1p6LuzoKR55_HgQJTWFInwj2FlIeo7_uAFXYuXe_Lw7Q/edit related to HA DNS Proxy CFP.

``` 
Need to determine L7 policy per-endpoint. Can we use epconfig.json to load per-endpoint policy, or do we need another discoverability mechanism?
This is not fast enough, but a PR to write epconfig.json on regeneration would be reasonable
```

**Comment by squeed** [comment]
@christarazi 
> What is the consequence if we don't update (2) on a periodic basis? [...] Does the list of ports + DNS server pods IPs (coredns) never get updated in the endpoint caches?

That's exactly right. This is caching the set of allowed IPs so we can correctly serve proxied DNS requests while the agent is starting up.

Since DNS rules are per port + destination, an endpoint could potentially have different DNS rules per destination. We also would like to be able to serve DNS requests very quickly when restarting the agent. Since a fresh agent will not know IP:Identity mappings, we cache IPs rather than identities. But then we never refresh the cache.

**Comment by squeed** [comment]
@vipul-21 have you had a chance to look at this? It would be a nice fix to have!

**Comment by vipul-21** [comment]
> @vipul-21 have you had a chance to look at this? It would be a nice fix to have!

@squeed  Sorry for the delay.(busy with some other stuff).
Yes, let me update this.

**Comment by vipul-21** [comment]
@squeed I updated the branch with the controller, but not seeing an option to reopen this PR ? Should I create a new one ?
